pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -524,3 +528,89 @@ func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
 
 	return r, true
 }
+
+// Implementation of sort interface for blockstorage version probing
+type APIVersionsByID []apiversions_v1.APIVersion
+
+func (apiVersions APIVersionsByID) Len() int {
+	return len(apiVersions)
+}
+
+func (apiVersions APIVersionsByID) Swap(i, j int) {
+	apiVersions[i], apiVersions[j] = apiVersions[j], apiVersions[i]
+}
+
+func (apiVersions APIVersionsByID) Less(i, j int) bool {
+	return apiVersions[i].ID > apiVersions[j].ID
+}
+
+func (os *OpenStack) Volumes() (IVolumes, error) {
+	switch os.bsOpts.BSVersion {
+	case "v1":
+		sClient, err := openstack.NewBlockStorageV1(os.provider, gophercloud.EndpointOpts{
+			Region: os.region,
+		})
+		if err != nil || sClient == nil {
+			glog.Errorf("Unable to initialize cinder client for region: %s", os.region)
+			return nil, err
+		}
+		return (IVolumes)(&VolumesV1{sClient, os.bsOpts}), nil
+	case "v2":
+		sClient, err := openstack.NewBlockStorageV2(os.provider, gophercloud.EndpointOpts{
+			Region: os.region,
+		})
+		if err != nil || sClient == nil {
+			glog.Errorf("Unable to initialize cinder v2 client for region: %s", os.region)
+			return nil, err
+		}
+		return (IVolumes)(&VolumesV2{sClient, os.bsOpts}), nil
+	case "auto":
+		sClient, err := openstack.NewBlockStorageV1(os.provider, gophercloud.EndpointOpts{
+			Region: os.region,
+		})
+		if err != nil || sClient == nil {
+			glog.Errorf("Unable to initialize cinder client for region: %s", os.region)
+			return nil, err
+		}
+
+		apiversions_v1.List(sClient).EachPage(func(page pagination.Page) (bool, error) {
```

Comment:
There is no v2 apiversions at all in the gophercloud library.
It is used _only_ to probe available versions. 
So it will work flawlessly if v1 is absent.